### ISSUE TITLE
hls4ml HGQ2 issues: softmax LUT mismatch + negative bitwidths

### ISSUE THREAD

**Issue by m-glowacki** [fastmachinelearning/hls4ml#1523]
# hls4ml HGQ2 issues: softmax LUT mismatch + negative bitwidths

## Summary

For HGQ2-quantized models (in this case using `QLinformerAttention` layers), **da4ml `comb_trace` is bit-accurate to Keras**, while **hls4ml diverges** due to two issues:

1. **Softmax LUT values** — hls4ml uses `std::exp()`/`1/x` at runtime instead of HGQ's trained tables
2. **Negative bitwidths** — `extract_fixed_quantizer_config` produces garbage types for learned zero-width channels

### Root cause

hls4ml's C++ softmax (`nnet_utils/nnet_activation.h`) populates LUT tables at first call:

```cpp
void init_exp_table(...) {
    for (i = 0; i < table_size; i++) {
        table_out[i] = std::exp(x);    // generic exp()
    }
}

void init_invert_table(...) {
    for (i = 0; i < table_size; i++) {
        table_out[i] = 1 / x;          // generic 1/x
    }
}
```

HGQ's `QSoftmax` uses `QUnaryFunctionLUT` layers (`exp_table`, `inv_table`) that store **trained LUT values** computed during HGQ training with quantization applied. hls4ml's `_bit_exact=True` correctly uses the table **sizes** and **types** from HGQ quantizer specs, but the **values** are generic `exp()`/`1/x`, not HGQ's trained weights.

This is explicitly documented in hls4ml source:
```python
# bit_exact.py:780-782
if not node.attributes.get('_bit_exact', False):
    # Softmax is not bit-exact by default
    warn(f'Softmax layer {node.name} is converted from a frontend '
         f'not supporting bit-exact softmax.')
```

###  first attention block (`q_softmax`, stable implementation)

| | HGQ (trained) | hls4ml  |
|---|---|---|
| Exp table (8 entries) | `[0.75, 0.75, 0.75, 0.75, 0.75, 0.25, 0.0, 0.0]` | different `exp(x * scale)` values |
| Inv table (512 entries) | mostly 0.0, upper block 0.5 | different `1/x` values |

## Per-output-dimension error (Keras vs hls4ml)

| Dim | Keras mean | HLS mean | MAE | corr |
|-----|-----------|----------|-----|------|
| 0 | 0.036 | 0.025 | 0.026 | 0.55 |
| 1 | 0.051 | 0.050 | 0.028 | 0.86 |
| 2 | -0.009 | -0.017 | 0.023 | 0.89 |
| 3 | -0.150 | -0.138 | 0.015 | 0.52 |
| 4 | -0.138 | -0.128 | 0.034 | 0.55 |
| 5 | 0.046 | 0.053 | 0.020 | 0.85 |
| 6 | 0.011 | 0.011 | 0.019 | 0.82 |
| 7 | 0.025 | 0.026 | 0.014 | 0.63 |
| 8 | -0.059 | -0.032 | 0.031 | 0.87 |
| 9 | 0.068 | 0.031 | 0.050 | 0.80 |
| 10 | -0.008 | -0.010 | 0.018 | 0.59 |
| 11 | 0.065 | 0.053 | 0.028 | 0.88 |
| 12 | 0.048 | 0.051 | 0.023 | 0.82 |
| 13 | -0.040 | -0.044 | 0.010 | 0.81 |
| 14 | 0.027 | 0.008 | 0.028 | 0.85 |
| 15 | -0.007 | 0.024 | 0.040 | 0.66 |

Manually widening the softmax output type from `ap_ufixed<1,0>` to `ap_ufixed<4,0>` (and downstream types to match attention1) produced **identical** error — the discrepancy is inside the LUT arithmetic before any truncation.

## Comparison plots

![](https://maciej.web.cern.ch/hls4ml/keras_vs_comb.png)

da4ml comb_trace matches Keras exactly (bit-accurate).

![](https://maciej.web.cern.ch/hls4ml/keras_vs_hls_hls4ml.png)

hls4ml diverges — scatter points deviate from the diagonal.

![](https://maciej.web.cern.ch/hls4ml/comb_vs_hls_hls4ml.png)

Direct da4ml vs hls4ml comparison shows the same discrepancy.

### Root cause

hls4ml's `extract_fixed_quantizer_config` (`hgq2/_base.py:19`) computes total bitwidth as `B = k + i + f` directly from HGQ quantizer specs:

```python
# hls4ml/converters/keras_v3/hgq2/_base.py:28-29
k, i, f = internal_q.kif
k, B, I = k, k + i + f, k + i   # B = total bits, I = integer bits
```

HGQ2 uses **heterogeneous quantizers** — per-element `(k, i, f)` where some channels learn **negative total bitwidths** (`k + i + f < 0`) as the training zeroes them out. When `B < 0`, the resulting `FixedPrecisionType(-5, 0)` produces garbage C++ types:

```
ap_fixed<-5,0>          ← negative width, template error
ap_ufixed<2,32>         ← 2 bits with 32 integer bits
```

This affected multiple layers including attention2's query/key input quantizers and the QK einsum output.

### Evidence (before fix)

```
attention2_query_iq_t:      ap_ufixed<2,32>     ← garbage
attention2_attention2_QK_t: ap_ufixed<2,32>     ← garbage
attention2_attention2_aV_t: ap_ufixed<1,0>      ← too narrow
```

### Fix applied

Clamp `B >= 1` and `I >= 0` in the extract function:

```python
B = np.clip(B, 1, None)
I = np.clip(I, 0, None)
```

Same fix for `fixed_quantizer_to_hls4ml_t` in the softmax handler. After fix:

```
attention2_query_iq_t:      ap_fixed<2,1>        ← correct
attention2_attention2_QK_t: ap_fixed<4,2>        ← correct
```
